### ISSUE TITLE
Sorting by two columns with the same name fails with ambiguous column error

### ISSUE THREAD

**Issue by camsaul** [metabase/metabase#6042]
For example if you create a query that has two `count` columns in the results and then click the column header to sort the resulting query will fail because the sort is ambiguous. This can be confusing because it's not obvious how to remove sorts in the query builder (you need to use the `...` menu even though you didn't need to use it to add the sort).

There's a couple things we could do to fix this:

1.  disallow sorts for ambiguous columns since we know they'll break
2.  try to prevent ambiguous columns from being created in the first place. e.g. by naming our columns `count` and `count_2` in the query instead of having `clojure.java.jdbc` do it for us in the results.
3.  let people manually specify column names, perhaps just reusing the existing UI for choosing which columns to display/etc.

**Comment by salsakran** [comment]
1 is problematic because of "sum" often being used multiple times for different columns.

Another thing is that we should probably allow user specific names for return columns. Right now we just have the sql operator, eg the below where we lose context of what "average" is, is kind of ghetto
![screen shot 2017-09-25 at 2 55 22 pm](https://user-images.githubusercontent.com/107439/30832874-a5294a46-a201-11e7-8d52-baea1341817a.png)

**Comment by M4zD4i** [comment]
Reproduced this issue. Resulting SQL query is really seems starnge:
![question metabase 2017-10-19 11-42-06](https://user-images.githubusercontent.com/1096609/31762188-177065a2-b4c3-11e7-828d-86696fab4d04.png)

**Comment by camsaul** [comment]
@M4zD4i what's "really strange" about that SQL query? 😃

**Comment by M4zD4i** [comment]
It's broken as we have: "SELECT ... as `sum`, ... as `sum` ... ORDER BY `sum`", so `sum` filed is ambiguous. May be it make sense to concat user selected field_name and alias_name here. For example: sum(`is_bid`) resulting "AS `sum_is_bid`"?

**Comment by yanivgold** [comment]
@salsakran, is there a fix planned to be out soon for this one? it's quite an important feature.
Greatest solution for this as I see it is having the ability to choose the naming given to a measure as the sort by option, and have that calc as the order by statement

**Comment by whitemark** [comment]
Hello, the problem is still there, it is a real issue, any plan to fix it? 

Thanks heaps!
Mark

**Comment by cmann50** [comment]
I am also hitting this issue
![image](https://user-images.githubusercontent.com/2318067/35473441-cc87acfa-034e-11e8-9570-18d6bcf40bb7.png)

Not sure if it would be easier to code, but perhaps we could leverage the custom field box.  Basically you just do =sum2 and alias the column.  Not super elegant like being able to click on the label and rename, but would use that existing UI box.

![image](https://user-images.githubusercontent.com/2318067/35473457-fa4315da-034e-11e8-8435-ea61e69fab5e.png)

**Comment by whitemark** [comment]
Hello

Actually you are quite right, I just noticed that you can click on "+" then "custom expression" 

![image](https://user-images.githubusercontent.com/13211752/35481697-b3c54fa4-047c-11e8-9331-614483501ff5.png)

That works as well

Still, would be easier if it could be seamless 

Thanks for pointing me to the right direction!

Mark

**Comment by cmann50** [comment]
Oh, nice, you solved it for me!  I was using the drop down on the far right under the _grouped by_ area and it wasn't working - that's why I suggested changing the code.  Then I realized from your screenshot there was a custom expressions under the _view_ area also, and doing it there works perfectly.  Agreed it might make sense to add a line to the documentation, or maybe just finding this conversation will be enough for people to figure out how to do it.

**Comment by cmann50** [comment]
So, this workaround does work if you choose custom expression SUM ().  You can rename the column.   But if you try to use CumulativeSum() it will not rename the column even though you type in a different column name.

![image](https://user-images.githubusercontent.com/2318067/35484440-c3fab6e8-041d-11e8-9cd9-150e4ee8741d.png)

![image](https://user-images.githubusercontent.com/2318067/35484463-1b2126d2-041e-11e8-9544-3d24641c5fea.png)

**Comment by alatui** [comment]
+1
Same issue

**Comment by TurkerTunali** [comment]
If we try to solve it by creating custom expression then we can't sort it by that column.

**Comment by RobinDarrah** [comment]
@camsaul Are there any plans for this to be milestoned? This bug has been around for a while now, and really limits metabase as a data exploration tool.

**Comment by rhnBen** [comment]
![image](https://user-images.githubusercontent.com/34618563/50593142-600ae000-0eac-11e9-964f-3e9d47908fca.png)

**Comment by tlrobinson** [comment]
I believe queries with the same aggregation names now (likely since we made sure column names were unique?) don't fail completely, however it always sorts the first of the aggregations with the matching name, e.x.

```json
{
  "aggregation": [
    ["sum", ["field-id", 1]],
    ["sum", ["field-id", 2]]
  ],
  "breakout: [
    ["field-id", 3]
  ],
  "order_by": [
    ["asc", ["aggregation", 1]]
  ]
}
```

will sort by the first aggregation (`["sum", ["field-id", 1]]`) even though it should sort by the second one (`["sum", ["field-id", 2]]`)

**Comment by flamber** [comment]
Related to #6239, which is about sorting/ordering Custom Expression - go and upvote by clicking :+1:

**Comment by SuneelPodapati** [comment]
Issue still exists in **0.32.4** with SQL Server data sources. Are there any plans for a fix soon?

![image](https://user-images.githubusercontent.com/10687003/56132074-b48f2280-5fa6-11e9-9b0d-fc2acbc422f3.png)

**Comment by flamber** [comment]
I'm not sure when this was fixed - think it was PR #9429 - but I can sort same-name columns in 0.32.9:
![image](https://user-images.githubusercontent.com/1447303/60999139-a63b7300-a317-11e9-8f4f-c3616ec3c766.png)

**Comment by flamber** [comment]
This was fixed in 0.33.0

**Comment by camsaul** [comment]
@SuneelPodapati why the 👎?
